### PR TITLE
feat(backup): Add restore indices and restore backup tasks

### PR DESCRIPTION
**Pull request by dexter-mh-lee**
Use the datahub-upgrade framework to restore indices and restore backup. 
- Restore indices: Get the latest version of each aspect from local db and produce mae events for each of them
- Restore backup: use EbeansBackupReader to get each row of the local db. Update local db and produce mae events. (For now only supports local parquet)

## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -61,6 +61,7 @@ project.ext.externalDependency = [
     'guice': 'com.google.inject:guice:4.2.2',
     'guava': 'com.google.guava:guava:27.0.1-jre',
     'h2': 'com.h2database:h2:1.4.196',
+    'hadoopClient': 'org.apache.hadoop:hadoop-client:3.1.0',
```
Comment (jjoyce0510) [inline build.gradle:64]: Just to confirm: Have we run any sec scans since adding these deps?

File: datahub-kubernetes/datahub/templates/datahub-upgrade/datahub-restore-indices-job-template.yml [hunk 2]
```diff
@@ -0,0 +1,119 @@
+{{- if .Values.datahubUpgrade.enabled -}}
```
Comment (jjoyce0510) [inline datahub-kubernetes/datahub/templates/datahub-upgrade/datahub-restore-indices-job-template.yml:1]: Is this correct? Why does this depend on datahubUpgrade flag?

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/UpgradeStep.java [hunk 3]
```diff
@@ -31,4 +32,10 @@ default boolean isOptional() {
     return false;
   }
 
+  /**
+   * Returns whether or not to skip the step based on the UpgradeContext
+   */
+  default boolean skip(UpgradeContext context) {
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/UpgradeStep.java:38]: Question: Why do we need this API if the step can simply check the context and no-op inside it's "execute" method?

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/common/steps/ClearGraphServiceStep.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  public String id() {
+    return "ClearGraphServiceStep";
+  }
+
+  @Override
+  public boolean skip(UpgradeContext context) {
+    if (_alwaysRun) {
+      return false;
+    }
+    if (context.parsedArgs().containsKey(NoCodeUpgrade.CLEAN_ARG_NAME)) {
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/common/steps/ClearGraphServiceStep.java:34]: Should we add a "cleanup has been requested" report line here?

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/impl/DefaultUpgradeManager.java [hunk 5]
```diff
@@ -52,30 +53,38 @@ private UpgradeResult executeInternal(UpgradeContext context) {
...
+      upgradeReport.addLine(
+          String.format(String.format("Executing Step %s/%s: %s...", i + 1, steps.size(), step.id()), upgrade.id()));
 
       final UpgradeStepResult stepResult = executeStepInternal(context, step);
       stepResults.add(stepResult);
 
       // Apply Actions
       if (UpgradeStepResult.Action.ABORT.equals(stepResult.action())) {
-        upgradeReport.addLine(String.format("Step with id %s requested an abort of the in-progress update. Aborting the upgrade...", step.id()));
+        upgradeReport.addLine(
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/impl/DefaultUpgradeManager.java:70]: Nice!

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/impl/DefaultUpgradeManager.java [hunk 6]
```diff
@@ -52,30 +53,38 @@ private UpgradeResult executeInternal(UpgradeContext context) {
...
+            String.format("Step with id %s requested an abort of the in-progress update. Aborting the upgrade...",
+                step.id()));
         return new DefaultUpgradeResult(UpgradeResult.Result.ABORTED, upgradeReport);
       }
 
       // Handle Results
       if (UpgradeStepResult.Result.FAILED.equals(stepResult.result())) {
         if (step.isOptional()) {
-          upgradeReport.addLine(String.format("Failed Step %s/%s: %s. Step marked as optional. Proceeding with upgrade...", i + 1, steps.size(), step.id()));
+          upgradeReport.addLine(
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/impl/DefaultUpgradeManager.java:79]: Awesome!!!!

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/impl/DefaultUpgradeManager.java [hunk 7]
```diff
@@ -98,26 +107,18 @@ private UpgradeStepResult executeStepInternal(UpgradeContext context, UpgradeSte
...
-            i,
-            step.id(),
-            e.toString()
-        ));
-        result = new DefaultUpgradeStepResult(
-            step.id(),
-            UpgradeStepResult.Result.FAILED);
+        context.report()
+            .addLine(
+                String.format("Caught exception during attempt %s of Step with id %s: %s", i, step.id(), e.toString()));
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/impl/DefaultUpgradeManager.java:120]: Thank you for making this class better!!!

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/nocode/NoCodeUpgrade.java [hunk 8]
```diff
@@ -54,15 +59,16 @@ public String id() {
...
+      final EntityClient entityClient) {
     final List<UpgradeStep> steps = new ArrayList<>();
     steps.add(new RemoveAspectV2TableStep(server));
     steps.add(new GMSQualificationStep());
     steps.add(new UpgradeQualificationStep(server));
     steps.add(new CreateAspectTableStep(server));
     steps.add(new IngestDataPlatformsStep(entityService));
     steps.add(new DataMigrationStep(server, entityService, entityRegistry));
-    steps.add(new GMSEnableWriteModeStep());
+    steps.add(new GMSEnableWriteModeStep(entityClient));
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/nocode/NoCodeUpgrade.java:71]: Do we ever disable write mode??

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/ClearAspectV2TableStep.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  public ClearAspectV2TableStep(final EbeanServer server) {
+    _server = server;
+  }
+
+  @Override
+  public String id() {
+    return "ClearAspectV2TableStep";
+  }
+
+  @Override
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/ClearAspectV2TableStep.java:28]: Just confirming: No skip state required here?

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/RestoreBackup.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    steps.add(new ClearSearchServiceStep(searchClient, true));
+    steps.add(new ClearGraphServiceStep(graphClient, true));
+    steps.add(new ClearAspectV2TableStep(server));
+    steps.add(new RestoreStorageStep(entityService, entityRegistry));
+    return steps;
+  }
+
+  @Override
+  public List<UpgradeCleanupStep> cleanupSteps() {
+    return ImmutableList.of();
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/RestoreBackup.java:55]: Minor nit: Collections.emptyList()

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/RestoreStorageStep.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+      context.report().addLine("Starting backup restore...");
+      int numRows = 0;
+      Optional<String> backupReaderName = context.parsedArgs().get("BACKUP_READER");
+      if (!backupReaderName.isPresent() || !_backupReaders.containsKey(backupReaderName.get())) {
+        context.report().addLine("BACKUP_READER is not set or is not valid");
+        return new DefaultUpgradeStepResult(id(), UpgradeStepResult.Result.FAILED);
+      }
+
+      EbeanAspectBackupIterator iterator = _backupReaders.get(backupReaderName.get()).getBackupIterator(context);
+      if (iterator == null) {
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/RestoreStorageStep.java:65]: It wasn't immediately obvious that the iterator could be null. Should the iterator have some method on it for this?

ie. iterator.hasData()?

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/BackupReader.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+
+/**
+ * Base interface for BackupReader used for creating the BackupIterator to retrieve EbeanAspectV2 object to be
+ * ingested back into GMS
+ */
+public interface BackupReader {
+  String getName();
+
+  EbeanAspectBackupIterator getBackupIterator(UpgradeContext context);
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/BackupReader.java:13]: Again, I wish this interface didn't have to know about UpgradeContext

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/ParquetEbeanAspectBackupIterator.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+import java.time.Instant;
+import java.util.Optional;
+import lombok.RequiredArgsConstructor;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.parquet.hadoop.ParquetReader;
+
+
+/**
+ * Iterator to retrieve EbeanAspectV2 objects from the ParquetReader
+ * Converts the avro GeneralRecord object into EbeanAspectV2
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/ParquetEbeanAspectBackupIterator.java:15]: minor nit: GeneralRecord -> GenericRecord

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/ParquetEbeanAspectBackupIterator.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    _parquetReader.close();
+  }
+
+  private EbeanAspectV2 convertRecord(GenericRecord record) {
+    if (record == null) {
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/ParquetEbeanAspectBackupIterator.java:41]: Case is already covered above

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/RestoreIndices.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    steps.add(new GMSQualificationStep());
+    steps.add(new ClearSearchServiceStep(searchService, false));
+    steps.add(new ClearGraphServiceStep(graphService, false));
+    steps.add(new SendMAEStep(server, entityService, entityRegistry));
+    return steps;
+  }
+
+  @Override
+  public List<UpgradeCleanupStep> cleanupSteps() {
+    return ImmutableList.of();
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/RestoreIndices.java:52]: minor nit: Collections.emptyList()

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/RestoreIndices.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+import com.linkedin.metadata.graph.GraphService;
+import com.linkedin.metadata.models.registry.EntityRegistry;
+import com.linkedin.metadata.search.SearchService;
+import io.ebean.EbeanServer;
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class RestoreIndices implements Upgrade {
+  public static final String BATCH_SIZE_ARG_NAME = "batchSize";
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/RestoreIndices.java:20]: Are these used?

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+    return 0;
+  }
+
+  @Override
+  public Function<UpgradeContext, UpgradeStepResult> executable() {
+    return (context) -> {
+
+      context.report().addLine("Sending MAE from local DB...");
+      final int rowCount = _server.find(EbeanAspectV2.class).where().eq(EbeanAspectV2.VERSION_COLUMN, 0).findCount();
+      context.report().addLine(String.format("Found %s rows in legacy aspects table", rowCount));
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java:54]: This is not the legacy table right?

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+
+      int totalRowsMigrated = 0;
+      int start = 0;
+      int count = getBatchSize(context.parsedArgs());
+      while (start < rowCount) {
+
+        context.report()
```
Comment (jjoyce0510) [inline datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java:61]: same

File: docker/datahub-upgrade/env/docker-without-neo4j.env [hunk 19]
```diff
@@ -0,0 +1,28 @@
+EBEAN_DATASOURCE_USERNAME=datahub
```
Comment (jjoyce0510) [inline docker/datahub-upgrade/env/docker-without-neo4j.env:1]: Should this file be called "datahub-without-neo4j.env"?

File: docs/how/restore-indices.md [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+
+## Docker-compose
+
+Run the following command from root to send MAE for each aspect in the Local DB.
+
+```
+./docker/datahub-upgrade/datahub-upgrade.sh -u RestoreIndices
+```
+
+If you need to clear the search and graph indices before restoring, add `-a clean` to the end of the command.
```
Comment (jjoyce0510) [inline docs/how/restore-indices.md:18]: Consider adding a note about the default env variables used and how to change them to suit your deployment

File: docs/how/restore-indices.md [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+
+## Kubernetes
+
+Run `kubectl get cronjobs` to see if the restoration job template has been deployed. If you see results like below, you
```
Comment (jjoyce0510) [inline docs/how/restore-indices.md:22]: Good stuf!!!

File: gms/factories/src/main/java/com/linkedin/gms/factory/search/SearchServiceFactory.java [hunk 22]
```diff
@@ -18,6 +19,7 @@
   private ElasticSearchService elasticSearchService;
 
   @Bean(name = "searchService")
+  @Primary
```
Comment (jjoyce0510) [inline gms/factories/src/main/java/com/linkedin/gms/factory/search/SearchServiceFactory.java:22]: just curious, why is this needed?

File: gms/impl/src/main/java/com/linkedin/metadata/resources/entity/EntityResource.java [hunk 23]
```diff
@@ -204,10 +205,10 @@
    */
   @Action(name = "setWritable")
   @Nonnull
-  public Task<Void> setWriteable() {
+  public Task<Void> setWriteable(@ActionParam(PARAM_VALUE) @Optional("true") @Nonnull Boolean value) {
```
Comment (jjoyce0510) [inline gms/impl/src/main/java/com/linkedin/metadata/resources/entity/EntityResource.java:208]: Thanks for making this change!

File: metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanAspectDao.java [hunk 24]
```diff
@@ -74,8 +74,12 @@ public EbeanAspectDao(@Nonnull final EbeanServer server) {
     _server = server;
   }
 
-  public void setWritable() {
-    _canWrite = true;
+  public void setWritable(boolean canWrite) {
+    _canWrite = canWrite;
+  }
+
+  public void disableWrite() {
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanAspectDao.java:81]: We can delete this now right?

File: metadata-io/src/main/java/com/linkedin/metadata/graph/elastic/ElasticSearchGraphService.java [hunk 25]
```diff
@@ -90,36 +93,25 @@ public void addEdge(@Nonnull final Edge edge) {
...
-  public List<String> findRelatedUrns(
-      @Nullable final String sourceType,
-      @Nonnull final Filter sourceEntityFilter,
-      @Nullable final String destinationType,
-      @Nonnull final Filter destinationEntityFilter,
-      @Nonnull final List<String> relationshipTypes,
-      @Nonnull final RelationshipFilter relationshipFilter,
-      final int offset,
-      final int count) {
+  public List<String> findRelatedUrns(@Nullable final String sourceType, @Nonnull final Filter sourceEntityFilter,
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/graph/elastic/ElasticSearchGraphService.java:96]: ughhhhh hate these reformattings

Comment (jjoyce0510) [reply metadata-io/src/main/java/com/linkedin/metadata/graph/elastic/ElasticSearchGraphService.java:96]: this is much less readable imo. maybe we need to change the styling rules

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/ElasticSearchService.java [hunk 26]
```diff
@@ -30,6 +30,10 @@ public void configure() {
     indexBuilders.buildAll();
   }
 
+  @Override
+  public void clear() {
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/ElasticSearchService.java:34]: Do we need to do something here?

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/UpgradeStep.java (hunk 3 above)
Comment (dexter-mh-lee) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/UpgradeStep.java:38]: Because I wanted to mention in the logs that it was skipped. 

Right now it says that DeleteAspectV2TableStep  ran successfully which to me implies that the table was deleted. But it actually wasn't, and was rather just skipped. I think it's good to make this explicit.

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/common/steps/ClearGraphServiceStep.java (hunk 4 above)
Comment (dexter-mh-lee) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/common/steps/ClearGraphServiceStep.java:34]: That log is added in the execute function. I can move it to here?

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/common/steps/GMSDisableWriteModeStep.java [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+import lombok.RequiredArgsConstructor;
+
+
+@RequiredArgsConstructor
+public class GMSDisableWriteModeStep implements UpgradeStep {
+  private final EntityClient _entityClient;
+
+  @Override
+  public String id() {
+    return "GMSEnableWriteModeStep";
```
[the remark this answers is not in the document]
Comment (dexter-mh-lee) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/common/steps/GMSDisableWriteModeStep.java:18]: OMG making this mistake way too many times

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/impl/DefaultUpgradeManager.java (hunk 5 above)
Comment (dexter-mh-lee) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/impl/DefaultUpgradeManager.java:70]: My formatter :(

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/nocode/NoCodeUpgrade.java (hunk 8 above)
Comment (dexter-mh-lee) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/nocode/NoCodeUpgrade.java:71]: During backup restore!

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/ClearAspectV2TableStep.java (hunk 9 above)
Comment (dexter-mh-lee) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/ClearAspectV2TableStep.java:28]: For now, I haven't seen any use case for it. The other ones were used by the RestoreIndices task which doesn't require clearing the MySQL table.

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/RestoreStorageStep.java (hunk 11 above)
Comment (dexter-mh-lee) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/RestoreStorageStep.java:65]: This is hard because there is no null iterator. I'll make this Optional to make this more explicit.

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/BackupReader.java (hunk 12 above)
Comment (dexter-mh-lee) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/BackupReader.java:13]: This was mainly because of the logging. If we can just log.info, then we don't need to pass in the context.

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/LocalParquetReader.java [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+      context.report().addLine("BACKUP_FILE_PATH must be set to run RestoreBackup through local parquet file");
+      return null;
+    }
+
+    try {
+      ParquetReader<GenericRecord> reader = AvroParquetReader.<GenericRecord>builder(new Path(path.get())).build();
+      return new ParquetEbeanAspectBackupIterator(reader);
+    } catch (IOException e) {
+      context.report().addLine(String.format("Failed to build ParquetReader: %s", e));
+      return null;
```
[the remark this answers is not in the document]
Comment (dexter-mh-lee) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/LocalParquetReader.java:44]: What happens when an exception is thrown here? Does it say the step failed?

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/ParquetEbeanAspectBackupIterator.java (hunk 13 above)
Comment (dexter-mh-lee) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/ParquetEbeanAspectBackupIterator.java:15]: done

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/ParquetEbeanAspectBackupIterator.java (hunk 14 above)
Comment (dexter-mh-lee) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/ParquetEbeanAspectBackupIterator.java:41]: good catch. removed

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java (hunk 17 above)
Comment (dexter-mh-lee) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java:54]: done

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java (hunk 18 above)
Comment (dexter-mh-lee) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java:61]: done

File: docs/how/restore-indices.md (hunk 20 above)
Comment (dexter-mh-lee) [reply docs/how/restore-indices.md:18]: Added a link to the datahub-upgrade readme section on env variables. Also updated that doc

File: gms/factories/src/main/java/com/linkedin/gms/factory/search/SearchServiceFactory.java (hunk 22 above)
Comment (dexter-mh-lee) [reply gms/factories/src/main/java/com/linkedin/gms/factory/search/SearchServiceFactory.java:22]: I am not sure :( Build failed because it said there were two instances of SearchService one from here and one from ElasticSearchServiceFactory. I have no idea why this is like this while graph service works fine.

File: metadata-io/src/main/java/com/linkedin/metadata/graph/elastic/ElasticSearchGraphService.java (hunk 25 above)
Comment (dexter-mh-lee) [reply metadata-io/src/main/java/com/linkedin/metadata/graph/elastic/ElasticSearchGraphService.java:96]: Me too :( I'm going to revert this file and add back the one function I added.

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/ElasticSearchService.java (hunk 26 above)
Comment (dexter-mh-lee) [reply metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/ElasticSearchService.java:36]: good catch

File: datahub-kubernetes/datahub/templates/datahub-upgrade/datahub-restore-indices-job-template.yml (hunk 2 above)
Comment (dexter-mh-lee) [reply datahub-kubernetes/datahub/templates/datahub-upgrade/datahub-restore-indices-job-template.yml:1]: All k8s files under datahub-upgrade are all enabled by the same flag

Comment (dexter-mh-lee) [reply datahub-kubernetes/datahub/templates/datahub-upgrade/datahub-restore-indices-job-template.yml:1]: Enabling it doesn't do much until you actually run the job since these cronjobs are suspended by default

File: build.gradle (hunk 1 above)
Comment (dexter-mh-lee) [reply build.gradle:65]: good point

Comment (dexter-mh-lee) [reply build.gradle:65]: looks good

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/UpgradeStep.java (hunk 3 above)
Comment (jjoyce0510) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/UpgradeStep.java:38]: okay makes sense.. agree its useful to know

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/common/steps/ClearGraphServiceStep.java (hunk 4 above)
Comment (jjoyce0510) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/common/steps/ClearGraphServiceStep.java:34]: no lets just keep that

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/RestoreStorageStep.java (hunk 11 above)
Comment (jjoyce0510) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/RestoreStorageStep.java]: ty

File: datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/BackupReader.java (hunk 12 above)
Comment (jjoyce0510) [reply datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/backupreader/BackupReader.java]: forgot about this -- we should be able to just use the logger as well

File: metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanAspectDao.java (hunk 24 above)
Comment (dexter-mh-lee) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanAspectDao.java]: done

**Review by jjoyce0510 — APPROVED** [review]
Overall, LGTM.

Thanks for addressing the comments!

**Review by shirshanka — APPROVED** [review]
LGTM!
